File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/stateengine/SeriesStack.java
package org.firstinspires.ftc.teamcode.teamcode.stateengine;


import org.firstinspires.ftc.teamcode.teamcode.hardware.RobotHardware;

public class SeriesStack extends StackList implements State {

    private int stackIndex = 0;


    @Override
    public void init(RobotHardware rh) {
        super.initStack(rh);
    }

    @Override
    public void run() {

        // go to next State if current State is done
        if (getStack()[stackIndex].isDone()) {
            stackIndex++;
        }

        // run State
        if (!isDone()) {
            getStack()[stackIndex].run();
        }
    }

    @Override
    public void stop() {

    }

    @Override
    public boolean isDone() {
        // when this stack has finished all States,
        // this stack will be done
        return stackIndex == getStack().length;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/stateengine/SeriesStack.kt
package org.firstinspires.ftc.teamcode.stateengine

import org.firstinspires.ftc.teamcode.hardware.RobotHardware

class SeriesStack(private val stack: Array<State>) : StackList(stack), State {
  private var stackIndex = 0

  override val isDone: Boolean
    get() = stackIndex == stack.size // When this stack has finished all States, this stack will be done

  override fun init(rh: RobotHardware) {
    super.initStack(rh)
  }

  override fun run() {
    // Go to next State if current State is done
    if (stack[stackIndex].isDone) {
      stackIndex++
    }

    // Run State
    if (!isDone) {
      stack[stackIndex].run()
    }
  }
}
